### ISSUE TITLE
JsonSerializer support for TimeSpan in 3.0?

### ISSUE THREAD

**Issue by martincostello** [dotnet/runtime#29932]
EDIT @eiriktsarpalis: See https://github.com/dotnet/runtime/issues/29932#issuecomment-857574430 for the API proposal.

_Apologies if this is already answered somewhere else, but if it was my search abilities failed me._

I'm trying out the preview 6 of 3.0 with some existing production application code, and one of the issues that's come out of testing is that some of our existing objects we use in our API contracts use properties which are `TimeSpan` values, which are then represented as strings.

Is support for `TimeSpan` properties planned for 3.0 for the new System.Text.Json APIs?

If it's not going to be that gives us notice to do some refactoring ahead of September to change them to strings so we can use the new serializer, where as if it's planned but just not implemented yet then we just need to wait for a later preview to get this working.

Below is a minimal repro unit test that demonstrates the failure for `TimeSpan` to be handled compared to our existing JSON .NET code.

```csharp
using System;
using System.Text.Json.Serialization;
using Xunit;
using JsonConvert = Newtonsoft.Json.JsonConvert;
using JsonSerializer = System.Text.Json.Serialization.JsonSerializer;

namespace JsonSerializerTimeSpanNotSupportedException
{
    public static class Repro
    {
        [Fact]
        public static void Can_Deserialize_Object_With_SystemTextJson()
        {
            // Arrange
            string json = "{\"child\":{\"value\":\"00:10:00\"}}";

            // Act (fails in preview 6, throws: System.Text.Json.JsonException : The JSON value could not be converted to System.TimeSpan. Path: $.child.value | LineNumber: 0 | BytePositionInLine: 28.)
            var actual = JsonSerializer.Parse<Parent>(json, new JsonSerializerOptions() { PropertyNamingPolicy = JsonNamingPolicy.CamelCase });

            // Assert
            Assert.NotNull(actual);
            Assert.NotNull(actual.Child);
            Assert.Equal(TimeSpan.FromMinutes(10), actual.Child.Value);
        }

        [Fact]
        public static void Can_Deserialize_Object_With_NewtonsoftJson()
        {
            // Arrange
            string json = "{\"child\":{\"value\":\"00:10:00\"}}";

            var actual = JsonConvert.DeserializeObject<Parent>(json);

            // Assert
            Assert.NotNull(actual);
            Assert.NotNull(actual.Child);
            Assert.Equal(TimeSpan.FromMinutes(10), actual.Child.Value);
        }

        private sealed class Parent
        {
            public Child Child { get; set; }
        }

        private sealed class Child
        {
            public TimeSpan Value { get; set; }
        }
    }
}
```

**Comment by alvipeo** [comment]
Oh, shit, this hurts!

**Comment by adams85** [comment]
> For anyone blocked by this, here's a NuGet package with a converter (JsonTimeSpanConverter) we can use ahead of the 5.0 drop: [Macross.Json.Extensions](https://www.nuget.org/packages/Macross.Json.Extensions/)
> 
> Supports TimeSpan & Nullable<TimeSpan> types.

According to my tests, nullable value types are already handled by the framework. You don't need much more than the following:

```csharp
public class DelegatedStringJsonConverter<T> : JsonConverter<T> where T : notnull
{
    private static readonly bool s_typeAllowsNull = !typeof(T).IsValueType || Nullable.GetUnderlyingType(typeof(T)) != null;

    private readonly Func<string, T> _parse;
    private readonly Func<T, string> _toString;

    public DelegatedStringJsonConverter(Func<string, T> parse, Func<T, string> toString)
    {
        _parse = parse;
        _toString = toString;
    }

    public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        if (reader.TokenType == JsonTokenType.Null)
            return s_typeAllowsNull ? default(T)! : throw new JsonException($"{typeof(T)} does not accept null values.");

        return _parse(reader.GetString());
    }

    public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
    {
        if (value is null)
        {
            writer.WriteNullValue();
            return;
        }

        writer.WriteStringValue(_toString(value));
    }
}
```

Then configure `JsonSerializerOptions` like this:

```csharp
options.Converters.Add(new DelegatedStringJsonConverter<TimeSpan>(
    value => TimeSpan.Parse(value, CultureInfo.InvariantCulture),
    value => value.ToString(null, CultureInfo.InvariantCulture)));
```

I don't think it's a good idea to add a 3rd party dependency if you can cope with such a few lines of extra code. We're not in NPM land after all. ;)

Of course, it'd be best if we didn't need a workaround for such a basic type at all.

**Comment by CodeBlanch** [comment]
@adams85 FYI Nullable value types are bugged < .NET 5 when used with JsonConverterAttribute. See #32006. I personally prefer to use the attribute style over global configuration, hence the extensions & bug fix, but your converter is excellent. Do you mind if I add it to Macross.Json.Extensions to help anyone else out that might also benefit from it and doesn't mind traveling to "NPM land" on occasion? :)

**Comment by adams85** [comment]
@CodeBlanch Thank you for pointing this quirk with `JsonConverterAttribute` out. It's definitely part of the full story.

And, of course, you are granted to add my converter to your lib if you like it. :)

**Comment by taspeotis** [comment]
Starting with .NET 5.0 Preview 5 I've been getting `Cannot skip tokens on partial JSON` errors round-tripping my entities to JSON with System.Text.Json. The offending line and character is the colon in `"Ticks":`

Anyway if I use [the workaround](https://github.com/dotnet/runtime/issues/29932#issuecomment-540200476) and serialize `TimeSpan` in a sensible fashion it works fine.

+1 from me because my initial reaction upon opening the `.json` file to inspect it only to find the `TimeSpan` serialized in all of its structural glory was "surely not..."

**Comment by layomia** [comment]
From @jsedlak in https://github.com/dotnet/runtime/issues/42356:

> # Issue Title
> The System.Text.Json.JsonSerializer class is unable to deserialize a TimeSpan property, even though it can serialize it.
> 
> # General
> Sample project demonstrating this issue is available here: https://github.com/jsedlak/TestTimeSpan
> 
> If you return a TimeSpan property in a WebApi, it is correctly serialized to JSON:
> 
> `[ { "forecastLength": { "ticks": 36000000000, "days": 0, "hours": 1, "milliseconds": 0, "minutes": 0, "seconds": 0, "totalDays": 0.041666666666666664, "totalHours": 1, "totalMilliseconds": 3600000, "totalMinutes": 60, "totalSeconds": 3600 } } ]`
> 
> But the default deserializer extensions (`HttpClient.GetFromJsonAsync`) cannot handle the property. It returns an empty TimeSpan. A custom converter must be used to deserialize the object.
> 
> # DotNet Info
> `.NET Core SDK (reflecting any global.json):
> Version: 3.1.302
> Commit: 41faccf259
> 
> Runtime Environment:
> OS Name: Windows
> OS Version: 10.0.20201
> OS Platform: Windows
> RID: win10-x64
> Base Path: C:\Program Files\dotnet\sdk\3.1.302\
> 
> Host (useful for support):
> Version: 3.1.6
> Commit: 3acd9b0cd1
> 
> .NET Core SDKs installed:
> 3.1.302 [C:\Program Files\dotnet\sdk]
> 
> .NET Core runtimes installed:
> Microsoft.AspNetCore.All 2.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
> Microsoft.AspNetCore.App 2.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
> Microsoft.AspNetCore.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
> Microsoft.NETCore.App 2.1.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
> Microsoft.NETCore.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
> Microsoft.WindowsDesktop.App 3.1.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]`

**Comment by layomia** [comment]
We'd like to address this in .NET 6.0 and would appreciate a community contribution here. The workaround in the meantime is to use a custom converter e.g. https://github.com/dotnet/runtime/issues/29932#issuecomment-540200476. The implementation should be based on the ISO8601 format for duration to be consistent with the behavior for `DateTime` and `DateTimeOffset`.

**Comment by taspeotis** [comment]
> We'd like to address this in .NET 6.0

.NET 5 isn't released yet, I'd love if we bit the bullet in v5 (semver permits major breaking changes v3 => v5) instead of waiting for v6.

**Comment by layomia** [comment]
We couldn't fit this feature into 5.0. We had to prioritize along with the rest of the [5.0 feature work](https://github.com/dotnet/runtime/issues/41313), and this one didn't fit. FWIW this was close to the cut line, thus the late move.

Just to highlight the simple workaround  with a [custom converter](https://docs.microsoft.com/dotnet/standard/serialization/system-text-json-converters-how-to) again:

```cs
public class TimeSpanConverter : JsonConverter<TimeSpan>
{
    public override TimeSpan Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return TimeSpan.Parse(reader.GetString());
    }

    public override void Write(Utf8JsonWriter writer, TimeSpan value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.ToString());
    }
}

var options = new JsonSerializerOptions { Converters = { new TimeSpanConverter() };
JsonSerializer.Serialize(myObj, options);

...
```

**Comment by YohDeadfall** [comment]
@layomia, I can take this for 6.0.

**Comment by BrunoZell** [comment]
I would recomment to use the invariant culture when opting to build a json converter yourself:

```csharp
public class TimeSpanConverter : JsonConverter<TimeSpan>
{
    public override TimeSpan Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        return TimeSpan.Parse(reader.GetString(), CultureInfo.InvariantCulture);
    }

    public override void Write(Utf8JsonWriter writer, TimeSpan value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.ToString(format: null, CultureInfo.InvariantCulture));
    }
}
```

**Comment by aquacat123** [comment]
I am working with asp.net core 3.1 and ran into this problem today.  I am using teleriks grid with signalr and inline editing on rows that contain nullable timespans and can see that the payload contains the timespans data but deserialization fails.  I ended up using the Macross.Json.Extensions library and reworking some of my javascript to make it work, not ideal though, i hope this gets a proper fix.

**Comment by tjipke** [comment]
And if you are looking for ISO8601 format, you can use this:

    public class TimeSpanConverter : JsonConverter<TimeSpan>
    {
        public override TimeSpan Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            var stringValue = reader.GetString();
            return System.Xml.XmlConvert.ToTimeSpan(stringValue); //8601
        }

        public override void Write(Utf8JsonWriter writer, TimeSpan value, JsonSerializerOptions options)
        {
            var stringValue = System.Xml.XmlConvert.ToString(value); //8601
            writer.WriteStringValue(stringValue);
        }
    }

**Comment by gojanpaolo** [comment]
Please consider `TimeSpan` of at least 24 hours.. e.g. `"24:00:00"` should be equivalent to `new TimeSpan(24, 0, 0)`

**Comment by CodeBlanch** [comment]
@gojanpaolo 

> Please consider TimeSpan of at least 24 hours.. e.g. "24:00:00" should be equivalent to new TimeSpan(24, 0, 0)

That isn't really a JSON thing, it's part of the TimeSpan parse logic. Short answer: Use "1.00:00:00" for 24 hours. Long answer: I went spelunking through the runtime code a while back to figure out why "24:00:00" doesn't parse as expected and [wrote a blog post about it.](https://blog.macrosssoftware.com/index.php/2020/06/04/timespan-parse240000-may-surprise-you/)

**Comment by AmrAlSayed0** [comment]
The ISO 8601 representation has a couple of possibly solvable problems. It is like this ```PnYnMnDTnHnMnS```.
1. It uses months and years which the TimeSpan doesn't use (I'm assuming because they aren't fixed periods of time).
2. The lowest unit is seconds which means you can't store exact number of ticks if you want.
BUT
...you could omit any part of that string (e.g P5D is a valid ISO duration) so while serializing you could not use the Y and M parts and stop at the D but that would not be an ISO duration because and I quote "individual date and time values cannot exceed their moduli ". Also the deserializer has to be prepared for this.
and
...decimal values are allowed meaning you could specify durations down to millisecond if you want (e.g P0.005S is a valid ISO duration of 5 milliseconds)

**Comment by zmhh** [comment]
If you are trying to read TimeSpan that was written with default System.Text.Json behavior, the touted workaround does not work.

Instead you can use this to read a TimeSpan written by the default System.Text.Json behavior:

```
public class DefaultTimespanReader : JsonConverter<TimeSpan>
    {
        public override TimeSpan Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            var ticksSpan = Encoding.UTF8.GetBytes(nameof(TimeSpan.Ticks)).AsSpan();
            TimeSpan? result = null;
            while (reader.TokenType != JsonTokenType.EndObject && reader.Read())
            {
                if (reader.TokenType == JsonTokenType.PropertyName && reader.ValueTextEquals(ticksSpan))
                {
                    reader.Read();
                    result = TimeSpan.FromTicks(long.Parse(Encoding.UTF8.GetString(reader.ValueSpan)));
                }
            }

            return result ?? throw new JsonException("Failed to deserialize timespan.");
        }

        public override void Write(Utf8JsonWriter writer, TimeSpan value, JsonSerializerOptions options)
        {
            throw new NotImplementedException("This is only meant for reading timespans written with default System.Text.Json behavior.");
        }
    }
```

**Comment by MartyIX** [comment]
@YohDeadfall Any PR to look forward to? :-)

**Comment by YohDeadfall** [comment]
Should be as soon as I end with naming policies, that story started before .NET 3.0, so I would like to complete it first. A good thing here is that it already works, but needs some benchmarks and a bit of tuning.

**Comment by danmoseley** [comment]
Hi @YohDeadfall any update to share ? 🙂 I know we've had internal asks for this too (hence the add to the new github project above) .

**Comment by YohDeadfall** [comment]
Hello, @danmoseley! Planned to spend upcoming weekend to move it forward and recheck that all things done. Meanwhile, I have another issue assigned to me on which I made progress and on which collected a lot information in past years, so it would be nice if someone from the team will take a look at #782 and dotnet/corefx#41354, and give some feedback.

**Comment by danmoseley** [comment]
@ericstj for above comment. I know the team is working hard on the source generator at the moment.

**Comment by ericstj** [comment]
At the moment the team is pretty heads-down on the source-generator. I expect folks to have more cycles starting next week.  I'll ping the linked issue.

**Comment by YohDeadfall** [comment]
@eiriktsarpalis, as I wrote in the PR, I feel burned out and resigning therefore. It would be better to not prevent others from taking this.

**Comment by eiriktsarpalis** [comment]
Take care @YohDeadfall, and thank you for working on the feature.

**Comment by CodeBlanch** [comment]
@eiriktsarpalis @layomia I'm happy to pick this up if that would be helpful.

**Comment by eiriktsarpalis** [comment]
Thanks @CodeBlanch, I've assigned the issue to you. @YohDeadfall's PR #51492 would be a good starting point, in particular the [conversation around the serialization format](https://github.com/dotnet/runtime/pull/51492#pullrequestreview-639978603).

**Comment by CodeBlanch** [comment]
@eiriktsarpalis I just read through the thread, it sounds like we want to go with [standard TimeSpan format](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-timespan-format-strings#the-general-short-g-format-specifier) instead of ISO8601 can you confirm that for me, and then I'll get started?

**Comment by mattjohnsonpint** [comment]
The issue with using the ISO860 period format is that it can represent values that don't fit cleanly into a `TimeSpan`.  So if one had in their JSON something like `"P3M"`  (3 months), the best one could do for a `TimeSpan` would be to use an *average* month duration, such as 3 months of 30 days each of exactly 24 hours long.  Of course, such values wouldn't survive a round trip either, because it could come out as `"P90D"` or `"P2160H"`, or a variety of other ways.

This is why in [Noda Time](https://nodatime.org) there are separate `Period` and `Duration` types.  `Period` uses the ISO8601 format and preserves the units of measure (even when ambiguous), while `Duration` resolves to a specific length of time and uses a format similar to `TimeSpan`'s default string format (except it separates days with a `:` instead of `.`).

FWIW, my advice is indeed to align to `TimeSpan`'s constant (`c`) format, `[-][d.]hh:mm:ss[.fffffff]` as described [here](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-timespan-format-strings#the-constant-c-format-specifier).

**Comment by eiriktsarpalis** [comment]
Yes, I think we go for the standard TimeSpan format and aim for roundtrip compatibility with Json.NET. cc @JamesNK who might have opinions on the matter.

**Comment by eiriktsarpalis** [comment]
Following #53539 this feature will be adding new APIs to facilitate source generation:

```csharp
namespace System.Text.Json.Serialization.Metadata
{
    public static class JsonMetadataServices
    {
        public static System.Text.Json.Serialization.JsonConverter<System.TimeSpan> TimeSpanConverter { get; }
    }
}
```

**Comment by KevinCathcart** [comment]
This will be added to the netstandard2.0 builds too, right? (Since it can be, as TimeSpan is not a new type.) Thus the additional discussions on #53539 with respect to that won’t apply here?

**Comment by terrajobst** [comment]
* We believe it's an acceptable breaking change even though we serialized `TimeSpan` before as a POCO. If developers don't want the new behavior, they can manually remove the `TimeSpan` from the `JsonSerializerOptions` instance. The source generator is new, so it's not a breaking change there regardless.
* There is no way to configure the emitted format of the the time span (we emit it as `days:hours:minute:seconds`). We could add an API to configure this later, but we don't believe it's needed.

```C#
namespace System.Text.Json.Serialization.Metadata
{
    public static class JsonMetadataServices
    {
        public static JsonConverter<TimeSpan> TimeSpanConverter { get; }
    }
}
```
